File: cpp/Petshop.cpp
// import library
#include <iostream>
#include <string>
#include <list>

using namespace std;

class Petshop{
    // Private atribut
    private:
        string id;
        string nama_produk;
        string kategori_produk;
        int harga_produk;


    public:
        // contructors
        Petshop(){
            this->id = "";
            this->nama_produk = "";
            this->kategori_produk = "";
            this->harga_produk = 0;
        }

        Petshop(string id, string nama_produk, string kategori_produk, int harga_produk){
            this->id = id;
            this->nama_produk = nama_produk;
            this->kategori_produk = kategori_produk;
            this->harga_produk = harga_produk;
        }

        // Getter
        string get_id(){
            return this->id;
        }
        string get_nama_produk(){
            return this->nama_produk;
        }
        string get_kategori_produk(){
            return this->kategori_produk;
        }
        int get_harga_produk(){
            return this->harga_produk;
        }

        // Setter
        void set_id(string id){
            this->id = id;
        }
        void set_nama_produk(string nama_produk){
            this->nama_produk = nama_produk;
        }
        void set_kategori_produk(string kategori_produk){
            this->kategori_produk = kategori_produk;
        }
        void set_harga_produk(int harga_produk){
            this->harga_produk = harga_produk;
        }

// Method

    // Method menambahkan produk
    void tambah_produk(list<Petshop>& daftar_produk) {
        string id, nama_produk, kategori_produk;
        int harga_produk;
        cout << "id, nama produk, kategori, harga :";
        cin >> id >> nama_produk >> kategori_produk >> harga_produk;
        // Memasukkan produk ke list daftar_produk dengan push_back
        daftar_produk.push_back(Petshop(id, nama_produk, kategori_produk, harga_produk));
    }

    // Method menampilkan semua produk
    void tampilkan_produk(list<Petshop>& daftar_produk) { // Parameter merujuk ke list daftar_produk sebagai referensinya
        if (daftar_produk.empty()) { // Jika list nya kosong
            cout << "Tidak ada produk\n";
        } 
        else { // Jika list tidak kosong, maka tampilkan datanya
            for (auto &produk : daftar_produk) { // Looping sebanyak data yang ada di list, dengan produk adalah referensi yang menunjek ke setiap bagian list daftar_produk
                cout << produk.id << " " 
                     << produk.nama_produk << " " 
                     << produk.kategori_produk << " " 
                     << produk.harga_produk << "\n";
            }
        }
    }

    // Method mencari produk berdasarkan ID
    void mencari_produk(list<Petshop>& daftar_produk) {
        string id;
        cout << "id yang ingin dicari: ";
        cin >> id;

        for (auto &produk : daftar_produk) { // Looping sebanyak data yang ada di list, dengan produk adalah referensi yang menunjek ke setiap bagian list daftar_produk
            if (produk.get_id() == id) { // Jika id produk sama dengan id yang dicari
                cout << "Produk ditemukan:\n";
                cout << produk.id << " " 
                     << produk.nama_produk << " " 
                     << produk.kategori_produk << " " 
                     << produk.harga_produk << "\n";
                return;
            }
        }
        cout << "Produk dengan ID " << id << " tidak ditemukan\n";
    }

    // Method menghapus produk berdasarkan ID
    void hapus_produk(list<Petshop>& daftar_produk) {
        string id;
        cout << "id produk yang ingin dihapus: ";
        cin >> id;

        auto it = daftar_produk.begin(); // Iterator yang menunjuk ke awal list daftar_produk
        while (it != daftar_produk.end()) { // Selama iterator tidak menunjuk ke akhir list
            if (it->get_id() == id) { // Jika id produk sama dengan id yang dicari
                it = daftar_produk.erase(it); // Hapus produk dari list
                cout << "Produk berhasil dihapus\n";
                return;
            } 
            else {
                ++it; // Pindah ke data selanjutnya
            }
        }
        cout << "Produk dengan ID " << id << " tidak ditemukan\n";
    }

    // Method mengupdate produk berdasarkan ID
    void update_produk(list<Petshop>& daftar_produk) {
        string id;
        cout << "id produk yang ingin diupdate: ";
        cin >> id;

        for (auto &produk : daftar_produk) { // Looping sebanyak data yang ada di list, dengan produk adalah referensi yang menunjek ke setiap bagian list daftar_produk
            if (produk.get_id() == id) { // Jika id produk sama dengan id yang dicari
                string nama_produk, kategori_produk;
                int harga_produk;

                cout << "nama produk, kategori, harga yang baru: ";
                cin >> nama_produk;
                cin >> kategori_produk;
                cin >> harga_produk;

                produk.set_nama_produk(nama_produk);
                produk.set_kategori_produk(kategori_produk);
                produk.set_harga_produk(harga_produk);

                cout << "Produk sudah diupdate\n";
                return;
            }
        }
        cout << "Produk dengan ID " << id << " tidak ditemukan\n";
    }

    // Destructor
    ~Petshop(){
        
    }
};

File: php/Petshop.php
<?php
session_start();

class Petshop {
    // private attributes
    private $id;
    private $namaProduk;
    private $kategoriProduk;
    private $harga;
    private $foto;

    // constructor
    public function __construct($id = '', $namaProduk = '', $kategoriProduk = '', $harga = '', $foto = '') {
        $this->id = $id;
        $this->namaProduk = $namaProduk;
        $this->kategoriProduk = $kategoriProduk;
        $this->harga = $harga;
        $this->foto = $foto;
    }

    // getter
    public function getId() {
        return $this->id;
    }
    public function getNamaProduk() {
        return $this->namaProduk;
    }
    public function getKategoriProduk() {
        return $this->kategoriProduk;
    }
    public function getHarga() {
        return $this->harga;
    }
    public function getFoto() {
        return $this->foto;
    }

    // setter
    public function setId($id) {
        $this->id = $id;
    }
    public function setNamaProduk($namaProduk) {
        $this->namaProduk = $namaProduk;
    }
    public function setKategoriProduk($kategoriProduk) {
        $this->kategoriProduk = $kategoriProduk;
    }
    public function setHarga($harga) {
        $this->harga = $harga;
    }
    public function setFoto($foto) {
        $this->foto = $foto;
    }

    // menyimpan data produk ke sesi / session
    public function save() {
        // disimpan ke session agar data tersimpan sementara
        // produk adalah array yang menyimpan data produk
        // mengecek apakah elemen produk dalam array sesi / session sudah ada dan tidak null
        if (!isset($_SESSION['produk'])) { // jika tidak ada, maka buat array kosong
            $_SESSION['produk'] = []; // membuat array kosong
        }

        // cek apakah id saat ini sudah ada dalam sesi 
        if (isset($_SESSION['produk'][$this->id])) {
            return false; // jika sudah ada maka false / tidak menyimpan produk
        }

        // simpan foto
        // memanggil fungsi upload foto, lalu $_FILES['foto'] adalah file yang diupload, dan $this->id adalah id produk
        $fotoPath = $this->uploadFoto($_FILES['foto'], $this->id); 
        $this->setFoto($fotoPath); // memanggil fungsi set foto untuk menyimpan path foto ke atribut foto

        $_SESSION['produk'][$this->id] = $this; // menyimpan data produk ke dalam array sesi / session
        // produk yang ditambahkan akan tersimpan dalam sesi (nama, kategori, harga, dan foto)
        return true;
    }

    // mengambil semua produk dari sesi / session
    public static function getAll() {
        // cek apakah $_SESSION['produk'] sudah ada atau belum
        if (isset($_SESSION['produk'])) {  // jika sudah ada
            return $_SESSION['produk']; // return daftar produk 
        }
        else { // jika belum ada
            return []; // return array kosong
        }
        
    }

    // mengupdate data produk di session
    public function update() {
        if (isset($_SESSION['produk'][$this->id])) { // mengecek apakah id produk sudah ada dalam sesi
            // cek apakah ada foto baru yang diupload
            // jika ada foto baru, hapus foto lama dan upload yang baru
            if (!empty($_FILES['foto']['name'])) {
                $oldFoto = $_SESSION['produk'][$this->id]->getFoto(); // path foto lama
                if (!empty($oldFoto) && file_exists($oldFoto)) { // jika foto lama tidak kosong dan file foto lama ada
                    unlink($oldFoto); // hapus foto lama
                }

                $fotoPath = $this->uploadFoto($_FILES['foto'], $this->id); // upload foto baru
                $this->setFoto($fotoPath); // simpan path foto baru
            }  
            else { // jika tidak ada foto baru, pakai foto lama
                $this->setFoto($_SESSION['produk'][$this->id]->getFoto()); // ambil path fotonya dari session 
            }

            $_SESSION['produk'][$this->id] = $this; // simpan perubahan ke dalam session
            return true; // return true jika berhasil update
        }
        return false; // return false jika produk tidak ada dalam sesi
    }

    // menghapus produk dan foto dari session
    // dibuat static karna diapus berdasarkan id, tidak perlu buat obj petshop
    public static function delete($id) {
        if (isset($_SESSION['produk'][$id])) { // cek apakah produk ada dalam session
            $foto = $_SESSION['produk'][$id]->getFoto(); // ambil path foto produk
            if (!empty($foto) && file_exists($foto)) { // jika path foto tidak kosong dan file foto ada
                unlink($foto); // hapus foto
            }
            unset($_SESSION['produk'][$id]); // hapus produk dari sesi
            return true; // return true jika berhasil hapus
        }
        return false; // return false jika produk tidak ada dalam sesi
    }

    // Mmngupload foto ke folder uploads
    private function uploadFoto($file, $id) {
        $targetDir = "uploads/"; // folder tujuan penymipanan foto
        if (!is_dir($targetDir)) { // jika folder tidak ada
            mkdir($targetDir, 0777, true); // buat folder baru, o777 adalah aksese untuk semua user
        }
        // menamai file
        $fileName = $id . "_" . basename($file["name"]); // basename($file["name"]) itu mengambil nama asli file yang diunggah, lalu ditambahkan id didepannya
        $targetFilePath = $targetDir . $fileName; // menyimpan path akhir file
        // menyimpan file ke folder uploads
        if (move_uploaded_file($file["tmp_name"], $targetFilePath)) { // memindahkan file dari tmp_name ke targetFilePath(uploads)
            return $targetFilePath; // return path file jika berhasil upload
        }
        return ""; // return kosong jika gagal upload
    }

    // mencari produk berdasarkan id
    public function cari($id) {
        // mengecek apakah ada produk dalam sesi
        if (!isset($_SESSION['produk'])) { // jika tidak ada produk dalam sesi
            return null; // return null
        }
    
        foreach ($_SESSION['produk'] as $produk) { // looping untuk mengecek tiap produk dalam sesi
            if ($produk->getId() == $id) { // jika id produk sesuai dengan id yang dicari
                return $produk; // return produk yang dicari
            }
        }
    
        return null; // return null jika tidak ada produk yang sesuai
    }
    
    
}
?>
